### PR TITLE
Reading the LE3 WL Two-Point Correlation Function (Xi +-) from FITS file

### PR DESCRIPTION
**Pull request by FelicitasKeil**
This is the pull request for the issue #4. 
It reads the \$\xi_\pm$ from the FITS file that Martin Kilbinger provided and builds on his code for reading the files. See this [issue](https://gitlab.euclid-sgs.uk/pf-ist-likelihood/likelihood-implementation/-/issues/1111) in the Cloe gitlab. I added a unit test checking the dictionary keys as in the other unit tests and it is passing locally.

I had some comments to the issue, for clarity I will paste them here again.

1. This reads the \$\xi_\pm$  for the 2 point correlation in WL. Martin Kilbinger also sent me a fits file for the COSEBIs for example. Is this all we want for now? The Cloe gitlab issue is titled "2PCF - WL", but the description also mentions other fits files.

2. Then I have a question related to the "philosophy" of euclidlib. For now, I am including the file `photo_data.py`in the `photo` folder written by Martin Kilbinger and Jingwei Wang who are credited in the same file. Does this work for euclidlib? Otherwise we could shorten this file, because not everything is used (yet) or integrate it somewhere else.

### CODE REVIEW

**Comment by FelicitasKeil** [comment]
I will have a look at the tests.

**Comment by gcanasherrera** [comment]
Hello @FelicitasKeil, thanks for all the effort and commitment on this. It is highly appreciated.

@ntessore and I have been brainstorming this morning about what's the best approach and philosophy to follow for `euclidlib`. We would like to maxime the contributions while minimizing the hussle of maintaining the code in the future (as will be used by `CLOE`, for instance). 

As this is our first official PR, it will also pose a learning process for both sides, so thanks in advance for your patience. Nicolas will take over as the reviewer and will suggest few changes to make, starting by:

- [ ]  Moving your reading routines to a new `.py` (suggestion: `_le3_2pcf_wl.py`)
- [ ]  Agree on the name of the functions

On my side, I will investigate if the current internal python dictionary used in the Cells can be easily adptable for the correlation function too.


Thanks again (to both!)

**Comment by FelicitasKeil** [comment]
Hello, that all sounds good to me.
I moved my code to a new file `_le3_2pcf_wl.py` as you suggested and also changed the unit tests accordingly. They are still passing locally.
I am also happy to change the function names if you prefer something different.

File: euclidlib/photo/__init__.py [hunk 1]
```diff
@@ -1,5 +1,6 @@
 __all__ = [
     "angular_power_spectra",
+    "two_point_correlation_xi",
```
Comment (ntessore) [inline euclidlib/photo/__init__.py:3]: This needs to match the actual function name below.

File: euclidlib/photo/_le3_2pcf_wl.py [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+
+if TYPE_CHECKING:
+    from collections.abc import Sequence
+    from typing import Any, TypeAlias
+
+
+# type alias
+_DictKey: "TypeAlias" = Union[str, int, Tuple["_DictKey", ...]]
+
+def xi_tpcf(
```
Comment (ntessore) [inline euclidlib/photo/_le3_2pcf_wl.py:17]: I think this needs a more expressive name, like how we decided on `angular_power_spectra()` and `mixing_matrices()`. See other comment.

File: euclidlib/photo/_le3_2pcf_wl.py [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+    path: str | PathLike[str],
+    *,
+    include: Sequence[tuple[Any, ...]] | None = None,
+    exclude: Sequence[tuple[Any, ...]] | None = None,
```
Comment (ntessore) [inline euclidlib/photo/_le3_2pcf_wl.py:21]: As far as I can tell, we are not actually including or excluding individual combinations below.

File: euclidlib/photo/_le3_2pcf_wl.py [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+) -> dict[_DictKey, NDArray[Any]]:
```
Comment (ntessore) [inline euclidlib/photo/_le3_2pcf_wl.py:22]: This annotation only matches if we are actually returning values in the `_DictKey` format, see below.

File: euclidlib/photo/_le3_2pcf_wl.py [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+
+    z_combinations = list(tpcf_le3._data[corr_str].keys())
+
+    # Get LE3 data
+    xis = {}
+    for component in ("+", "-"):
+        xis[component] = {}
+        for i, z_comb in enumerate(z_combinations):
+            xis[component][z_comb] = tpcf_le3.get_xi_component(component,
+                                                               z_comb)
```
Comment (ntessore) [inline euclidlib/photo/_le3_2pcf_wl.py:43]: We are only loading the xi correlation functions here -- could we load w(θ) or γ(θ) at the same time? Or are they provided in different files?

Can we load and return `THETA`, `THMIN`, `THMAX`, weight for each angular bin?

File: euclidlib/photo/_le3_2pcf_wl.py [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+    return xis
```
Comment (ntessore) [inline euclidlib/photo/_le3_2pcf_wl.py:44]: Missing newline at end of file, we can reformat this before merging.

File: euclidlib/photo/photo_data.py
Comment (ntessore) [inline euclidlib/photo/photo_data.py:1]: There is a lot in this implementation we don't need. We should consider trimming it down to the essentials in a second step.

File: pyproject.toml [hunk 7]
```diff
@@ -25,6 +25,7 @@ dynamic = ["version"]
 dependencies = [
     "fitsio",
     "numpy",
+    "astropy",
```
Comment (ntessore) [inline pyproject.toml:28]: I'm not entirely happy with the dependency on astropy, since astropy routinely breaks projects. Maybe we can refactor the FITS reading to use fitsio later.

File: tests/photo/data/two_point_correlation_function_Xi.fits
Comment (ntessore) [inline tests/photo/data/two_point_correlation_function_Xi.fits:1]: Thanks for providing the example file!

**Review by ntessore — CHANGES_REQUESTED** [review]
Hi @FelicitasKeil, great work, particularly making the existing code work so well with euclidlib.

The only real change that we should consider is whether this function could be called `correlation_functions()` and be used to load not only xi^± but also w(θ) and γ(theta). Even if those two products are provided in separate files, would it be possible for a single `correlation_functions()` implementation to detect which kind of FITS was provided? Because from a cursory look, the reader class looks to be the same for all.

File: euclidlib/photo/_le3_2pcf_wl.py (hunk 5 above)
Comment (FelicitasKeil) [reply euclidlib/photo/_le3_2pcf_wl.py]: Hello, in the new commit, I added THMIN, THMAX on the highest level of the dictionary. Theta is added as an array also on the highest level. Then the weight(theta) is loaded for each redshift combination. I hope that is what you had in mind. The unit test is updated as well.
For the other correlation functions, I don't have the FITS files for them so I could not test that directly. But we could contact Martin Kilbinger or someone else for them.

File: euclidlib/photo/_le3_2pcf_wl.py (hunk 6 above)
Comment (FelicitasKeil) [reply euclidlib/photo/_le3_2pcf_wl.py]: Done.

File: euclidlib/photo/_le3_2pcf_wl.py (hunk 3 above)
Comment (FelicitasKeil) [reply euclidlib/photo/_le3_2pcf_wl.py]: I removed these lines now.

File: euclidlib/photo/_le3_2pcf_wl.py (hunk 2 above)
Comment (FelicitasKeil) [reply euclidlib/photo/_le3_2pcf_wl.py]: I changed it to `two_point_correlation`, since we are planning to read the other FITS files as well. Let me know if you prefer something different.

File: euclidlib/photo/__init__.py (hunk 1 above)
Comment (FelicitasKeil) [reply euclidlib/photo/__init__.py]: Thank you for noticing, it's matching now.

File: euclidlib/photo/_le3_2pcf_wl.py (hunk 4 above)
Comment (FelicitasKeil) [reply euclidlib/photo/_le3_2pcf_wl.py]: I'm sorry, I'm not familiar with the `_DictKey` format and could not really find documentation of it. Could you please tell me where to read about it or how to change my dictionary?

File: pyproject.toml (hunk 7 above)
Comment (FelicitasKeil) [reply pyproject.toml:28]: I am reading the new entries with fitsio but left the old ones with astropy for now.

**Comment by FelicitasKeil** [comment]
Hello @ntessore, thank you for the review. I implemented some changes (see other comments). For the other correlation functions and the corresponding FITS files, I don't have them, but we could ask for them (which might take some time). 
If we change from astropy to fitsio, we might not need the `photo_data.py` file any more. Let me know if we should have a new pull request for that task.

**Comment by martinkilbinger** [comment]
Hello,

The code to read gamma_t and w(theta) is implemented in the Tpcf_Reader class, lines 561 and 568.
@FelicitasKeil I can send you the gamma_t file, I have not yet run w(theta), but will hopefully do this in the next few weeks.

Happy that my code is being used! I don't know how the output of two_point_correlation() is being used further in the pipeline, I assume the IST:L.

But if I may provide some comments, I would recommend to use a class and not a dictionary. Some, I believe, useful functionality is implemented in the classes in photo_data.py, and I am happy to discuss this further. For example, it might be useful to have class functions that return xi+ and xi- independently, or jointly, in a consistent way with corresponding angular scales, depending of whether this is used for plotting or in a 3x2pt data vector. Similarly for E- and B-modes (some but not all estimators have BB). Other checks of metadata contained in the FITS file might be useful as well, e.g. checks of the available redshift combinations, consistency of angular scales between different estimators and the covariance, etc.

> Hello @ntessore, thank you for the review. I implemented some changes (see other comments). For the other correlation functions and the corresponding FITS files, I don't have them, but we could ask for them (which might take some time). If we change from astropy to fitsio, we might not need the `photo_data.py` file any more. Let me know if we should have a new pull request for that task.

**Comment by FelicitasKeil** [comment]
Thank you very much for your comments.
Exactly, this task was actually started from the IST:L side where I will continue the implementation and call the functions from `euclidlib`.

1. I am happy to change it to a class. However, the other functions, like the angular power spectra, return dictionaries, so I think it should be consistent. Maybe @ntessore or @gcanasherrera can give their opinion on this.

2. I can implement a choice of returning $\xi_+$, $\xi_-$ or both either in the dictionary or with class functions.

3. The checks, if all redshift combinations are available and if the angular scales are consistent, are a good idea as well.

I will wait with some of these changes until we decide if we switch to the class structure, because otherwise it would need to be changed again.

**Comment by ntessore** [comment]
Hi @FelicitasKeil! I'm sorry for taking so long to reply. I think the PR is in a really good shape. We had the idea to just merge code more or less as it is, and refactor things at a later stage, when we have amassed a good collection of functionality. So I don't think you need to worry about functions vs classes at this stage.

So I would suggest that

1. Let's leave it as it is for now, and refactor later if necessary.
2. I think returning both in the dictionary is good, the user can grab whatever they need. We can make the interface more flexible later, if we see a need for it.
3. Along the same nice, this will be really cool to have in the future, but let's just get whatever we have available right now merged.

To summarise: Let's get something going first, with the understanding that it will be iterated on later.

**Comment by FelicitasKeil** [comment]
Hello @ntessore, thank you for your reply. 
I removed the `_DictKey` format for now. I have the impression that it only makes sense if the dictionary has the same "depth" everywhere. This is no longer the case because I added `TH_MIN` and `TH_MAX` at the top level.
I also changed the function name to `correlation_functions` as you suggested earlier.
As I said, the style tests are not passing due to the added file `photo_data.py`.
Let me know if you meant anything else with leaving the code more or less as it is.

**Comment by ntessore** [comment]
Hi @FelicitasKeil! I think we should merge this, we can iterate later. @gcanasherrera agree?
